In [1]:
import sys
from csv import reader
from operator import add
import numpy as np
import re

In [2]:
crimefile = sc.textFile("/Users/guozhiqi-seven/Google Drive/NYU Master/Big Data/project/NYPD_Complaint_Data_Historic.csv")
crimefile.count() 

5101232

* There are totally 5101232 recorded crime complaints in the dataset.

In [3]:
crimefile.take(2)

['CMPLNT_NUM,CMPLNT_FR_DT,CMPLNT_FR_TM,CMPLNT_TO_DT,CMPLNT_TO_TM,RPT_DT,KY_CD,OFNS_DESC,PD_CD,PD_DESC,CRM_ATPT_CPTD_CD,LAW_CAT_CD,JURIS_DESC,BORO_NM,ADDR_PCT_CD,LOC_OF_OCCUR_DESC,PREM_TYP_DESC,PARKS_NM,HADEVELOPT,X_COORD_CD,Y_COORD_CD,Latitude,Longitude,Lat_Lon',
 '101109527,12/31/2015,23:45:00,,,12/31/2015,113,FORGERY,729,"FORGERY,ETC.,UNCLASSIFIED-FELO",COMPLETED,FELONY,N.Y. POLICE DEPT,BRONX,44,INSIDE,BAR/NIGHT CLUB,,,1007314,241257,40.828848333,-73.916661142,"(40.828848333, -73.916661142)"']

In [4]:
from pyspark.sql import SQLContext
import pandas as pd
#sc = SparkContext('local','example')  # if using locally
sql_sc = SQLContext(sc)

pandas_df = pd.read_csv("/Users/guozhiqi-seven/Google Drive/NYU Master/Big Data/project/NYPD_Complaint_Data_Historic.csv")

//anaconda/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [5]:
pandas_df.head()

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM      RPT_DT  \
0   101109527   12/31/2015     23:45:00          NaN          NaN  12/31/2015   
1   153401121   12/31/2015     23:36:00          NaN          NaN  12/31/2015   
2   569369778   12/31/2015     23:30:00          NaN          NaN  12/31/2015   
3   968417082   12/31/2015     23:30:00          NaN          NaN  12/31/2015   
4   641637920   12/31/2015     23:25:00   12/31/2015     23:30:00  12/31/2015   

   KY_CD                        OFNS_DESC  PD_CD  \
0    113                          FORGERY  729.0   
1    101  MURDER & NON-NEGL. MANSLAUGHTER    NaN   
2    117                  DANGEROUS DRUGS  503.0   
3    344     ASSAULT 3 & RELATED OFFENSES  101.0   
4    344     ASSAULT 3 & RELATED OFFENSES  101.0   

                          PD_DESC              ...               ADDR_PCT_CD  \
0  FORGERY,ETC.,UNCLASSIFIED-FELO              ...                      44.0   
1                             NaN              ...                     103.0   
2  CONTROLLED SUBSTANCE,INTENT TO              ...                      28.0   
3                       ASSAULT 3              ...                     105.0   
4                       ASSAULT 3              ...                      13.0   

  LOC_OF_OCCUR_DESC    PREM_TYP_DESC PARKS_NM  HADEVELOPT X_COORD_CD  \
0            INSIDE   BAR/NIGHT CLUB      NaN         NaN  1007314.0   
1           OUTSIDE              NaN      NaN         NaN  1043991.0   
2               NaN            OTHER      NaN         NaN   999463.0   
3            INSIDE  RESIDENCE-HOUSE      NaN         NaN  1060183.0   
4          FRONT OF            OTHER      NaN         NaN   987606.0   

  Y_COORD_CD   Latitude  Longitude                        Lat_Lon  
0   241257.0  40.828848 -73.916661  (40.828848333, -73.916661142)  
1   193406.0  40.697338 -73.784557  (40.697338138, -73.784556739)  
2   231690.0  40.802607 -73.945052  (40.802606608, -73.945051911)  
3   177862.0  40.654549 -73.726339  (40.654549444, -73.726338791)  
4   208148.0  40.738002 -73.987891     (40.7380024, -73.98789129)  

[5 rows x 24 columns]

In [6]:
crime_df = sqlContext.read.csv("/Users/guozhiqi-seven/Google Drive/NYU Master/Big Data/project/NYPD_Complaint_Data_Historic.csv",header=True)


In [7]:
crime_df

DataFrame[CMPLNT_NUM: string, CMPLNT_FR_DT: string, CMPLNT_FR_TM: string, CMPLNT_TO_DT: string, CMPLNT_TO_TM: string, RPT_DT: string, KY_CD: string, OFNS_DESC: string, PD_CD: string, PD_DESC: string, CRM_ATPT_CPTD_CD: string, LAW_CAT_CD: string, JURIS_DESC: string, BORO_NM: string, ADDR_PCT_CD: string, LOC_OF_OCCUR_DESC: string, PREM_TYP_DESC: string, PARKS_NM: string, HADEVELOPT: string, X_COORD_CD: string, Y_COORD_CD: string, Latitude: string, Longitude: string, Lat_Lon: string]

In [8]:
from csv import reader
crimeline = sc.textFile("/Users/guozhiqi-seven/Google Drive/NYU Master/Big Data/project/NYPD_Complaint_Data_Historic.csv")
header = crimeline.first() #header
crimeline=crimeline.filter(lambda line: line != header)
crimeline = crimeline.mapPartitions(lambda x: reader(x)).map(lambda x: (x[0],1))

crimeline = crimeline.map(lambda x: ('NotNaN',1) if x[0] != None else ('NaN',1))

In [9]:
crimeline.take(5) 

[('NotNaN', 1), ('NotNaN', 1), ('NotNaN', 1), ('NotNaN', 1), ('NotNaN', 1)]

In [10]:
header

'CMPLNT_NUM,CMPLNT_FR_DT,CMPLNT_FR_TM,CMPLNT_TO_DT,CMPLNT_TO_TM,RPT_DT,KY_CD,OFNS_DESC,PD_CD,PD_DESC,CRM_ATPT_CPTD_CD,LAW_CAT_CD,JURIS_DESC,BORO_NM,ADDR_PCT_CD,LOC_OF_OCCUR_DESC,PREM_TYP_DESC,PARKS_NM,HADEVELOPT,X_COORD_CD,Y_COORD_CD,Latitude,Longitude,Lat_Lon'

In [11]:
crimeline=crimeline.reduceByKey(add).collect()

In [12]:
crimeline

[('NotNaN', 5101231)]

In [17]:
crimeline = sc.textFile("/Users/guozhiqi-seven/Google Drive/NYU Master/Big Data/project/NYPD_Complaint_Data_Historic.csv")
header = crimeline.first() #header
crimeline=crimeline.filter(lambda line: line != header)
CMPLNT_TO_DT = crimeline.mapPartitions(lambda x: reader(x)).map(lambda x: (x[3],1))

CMPLNT_TO_DT = CMPLNT_TO_DT.map(lambda x: ('NotNaN',1) if x[0] != np.nan else ('NaN',1))

In [18]:
CMPLNT_TO_DT=CMPLNT_TO_DT.reduceByKey(add).collect()

In [19]:
CMPLNT_TO_DT

[('NotNaN', 5101231)]

In [28]:
from pyspark.sql.functions import col, count, sum

def count_null(c):
    """Use conversion between boolean and integer
    - False -> 0
    - True ->  1
    """
    return sum(col(c).isNull().cast("integer").alias(c))

exprs = [count_null(c) for c in crime_df.columns]
crime_df.agg(*exprs) 

DataFrame[sum(CAST((CMPLNT_NUM IS NULL) AS INT) AS `CMPLNT_NUM`): bigint, sum(CAST((CMPLNT_FR_DT IS NULL) AS INT) AS `CMPLNT_FR_DT`): bigint, sum(CAST((CMPLNT_FR_TM IS NULL) AS INT) AS `CMPLNT_FR_TM`): bigint, sum(CAST((CMPLNT_TO_DT IS NULL) AS INT) AS `CMPLNT_TO_DT`): bigint, sum(CAST((CMPLNT_TO_TM IS NULL) AS INT) AS `CMPLNT_TO_TM`): bigint, sum(CAST((RPT_DT IS NULL) AS INT) AS `RPT_DT`): bigint, sum(CAST((KY_CD IS NULL) AS INT) AS `KY_CD`): bigint, sum(CAST((OFNS_DESC IS NULL) AS INT) AS `OFNS_DESC`): bigint, sum(CAST((PD_CD IS NULL) AS INT) AS `PD_CD`): bigint, sum(CAST((PD_DESC IS NULL) AS INT) AS `PD_DESC`): bigint, sum(CAST((CRM_ATPT_CPTD_CD IS NULL) AS INT) AS `CRM_ATPT_CPTD_CD`): bigint, sum(CAST((LAW_CAT_CD IS NULL) AS INT) AS `LAW_CAT_CD`): bigint, sum(CAST((JURIS_DESC IS NULL) AS INT) AS `JURIS_DESC`): bigint, sum(CAST((BORO_NM IS NULL) AS INT) AS `BORO_NM`): bigint, sum(CAST((ADDR_PCT_CD IS NULL) AS INT) AS `ADDR_PCT_CD`): bigint, sum(CAST((LOC_OF_OCCUR_DESC IS NULL) AS IN

In [37]:
crime_df.agg(*exprs).toPandas()

sum(CAST((CMPLNT_NUM IS NULL) AS INT) AS `CMPLNT_NUM`)  \
0                                                  0        

   sum(CAST((CMPLNT_FR_DT IS NULL) AS INT) AS `CMPLNT_FR_DT`)  \
0                                                655            

   sum(CAST((CMPLNT_FR_TM IS NULL) AS INT) AS `CMPLNT_FR_TM`)  \
0                                                 48            

   sum(CAST((CMPLNT_TO_DT IS NULL) AS INT) AS `CMPLNT_TO_DT`)  \
0                                            1391478            

   sum(CAST((CMPLNT_TO_TM IS NULL) AS INT) AS `CMPLNT_TO_TM`)  \
0                                            1387785            

   sum(CAST((RPT_DT IS NULL) AS INT) AS `RPT_DT`)  \
0                                               0   

   sum(CAST((KY_CD IS NULL) AS INT) AS `KY_CD`)  \
0                                             0   

   sum(CAST((OFNS_DESC IS NULL) AS INT) AS `OFNS_DESC`)  \
0                                              18840      

   sum(CAST((PD_CD IS NULL) AS INT) AS `PD_CD`)  \
0                                          4574   

   sum(CAST((PD_DESC IS NULL) AS INT) AS `PD_DESC`)  \
0                                              4574   

                         ...                         \
0                        ...                          

   sum(CAST((ADDR_PCT_CD IS NULL) AS INT) AS `ADDR_PCT_CD`)  \
0                                                390          

   sum(CAST((LOC_OF_OCCUR_DESC IS NULL) AS INT) AS `LOC_OF_OCCUR_DESC`)  \
0                                            1127128                      

   sum(CAST((PREM_TYP_DESC IS NULL) AS INT) AS `PREM_TYP_DESC`)  \
0                                              33279              

   sum(CAST((PARKS_NM IS NULL) AS INT) AS `PARKS_NM`)  \
0                                            5093632    

   sum(CAST((HADEVELOPT IS NULL) AS INT) AS `HADEVELOPT`)  \
0                                            4848026        

   sum(CAST((X_COORD_CD IS NULL) AS INT) AS `X_COORD_CD`)  \
0                                             188146        

   sum(CAST((Y_COORD_CD IS NULL) AS INT) AS `Y_COORD_CD`)  \
0                                             188146        

   sum(CAST((Latitude IS NULL) AS INT) AS `Latitude`)  \
0                                             188146    

   sum(CAST((Longitude IS NULL) AS INT) AS `Longitude`)  \
0                                             188146      

   sum(CAST((Lat_Lon IS NULL) AS INT) AS `Lat_Lon`)  
0                                            188146  

[1 rows x 24 columns]

In [2]:
crimedata = sc.textFile("/Users/guozhiqi-seven/Google Drive/NYU Master/Big Data/project/NYPD_Complaint_Data_Historic.csv")
#crimedata.count() 

CMPLNT_FR_DT_year = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[1][-4:],1))
CMPLNT_FR_DT_year.reduceByKey(add).collect()

[('2014', 490363),
 ('', 655),
 ('2003', 1535),
 ('1971', 3),
 ('1966', 26),
 ('1948', 1),
 ('1900', 5),
 ('1942', 2),
 ('1961', 1),
 ('2008', 528675),
 ('1915', 7),
 ('2010', 509725),
 ('1996', 120),
 ('1991', 27),
 ('1975', 8),
 ('1941', 2),
 ('1965', 5),
 ('1909', 3),
 ('1998', 216),
 ('1970', 6),
 ('1995', 73),
 ('1969', 6),
 ('1955', 3),
 ('1920', 6),
 ('1979', 9),
 ('1905', 2),
 ('1986', 27),
 ('1908', 3),
 ('1906', 1),
 ('1997', 131),
 ('2005', 10767),
 ('1950', 3),
 ('1976', 3),
 ('1973', 7),
 ('1999', 334),
 ('1940', 1),
 ('2002', 1039),
 ('1913', 9),
 ('1994', 63),
 ('1946', 1),
 ('1993', 42),
 ('2012', 504128),
 ('2015', 468576),
 ('1982', 7),
 ('2001', 986),
 ('1938', 1),
 ('1990', 34),
 ('1959', 2),
 ('1914', 10),
 ('1960', 8),
 ('1910', 9),
 ('1980', 13),
 ('1954', 2),
 ('1919', 1),
 ('1962', 3),
 ('1992', 44),
 ('2004', 2095),
 ('1968', 10),
 ('2000', 902),
 ('1911', 7),
 ('2011', 498198),
 ('1978', 5),
 ('1974', 9),
 ('1956', 1),
 ('2006', 539024),
 ('2009', 510946),
 (

In [60]:
CMPLNT_FR_DT_month = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[1][:2],1))
CMPLNT_FR_DT_month.reduceByKey(add).collect() 

[('', 655),
 ('05', 447946),
 ('CM', 1),
 ('01', 416073),
 ('12', 396451),
 ('09', 437906),
 ('11', 404670),
 ('08', 455531),
 ('06', 440047),
 ('07', 455911),
 ('03', 421703),
 ('04', 415965),
 ('10', 445188),
 ('02', 363185)]

In [61]:
CMPLNT_FR_DT_date = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[1][3:5],1))
CMPLNT_FR_DT_date.reduceByKey(add).collect()

[('', 655),
 ('17', 167483),
 ('05', 169483),
 ('16', 165424),
 ('14', 166979),
 ('24', 163435),
 ('20', 173893),
 ('22', 165593),
 ('01', 212054),
 ('26', 160702),
 ('19', 164112),
 ('12', 168548),
 ('09', 166012),
 ('11', 166249),
 ('18', 166818),
 ('28', 164046),
 ('LN', 1),
 ('23', 165317),
 ('08', 164996),
 ('06', 165261),
 ('13', 166491),
 ('07', 165986),
 ('03', 165137),
 ('29', 150850),
 ('31', 94513),
 ('30', 150996),
 ('04', 165540),
 ('15', 175027),
 ('27', 163069),
 ('10', 171303),
 ('25', 162778),
 ('02', 165517),
 ('21', 166964)]

In [62]:
CMPLNT_FR_TM_hour = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[2][:2],1))
CMPLNT_FR_TM_hour.reduceByKey(add).collect()

[('', 48),
 ('17', 291049),
 ('05', 72068),
 ('CM', 1),
 ('16', 291289),
 ('14', 271976),
 ('20', 290218),
 ('24', 903),
 ('22', 251364),
 ('00', 242359),
 ('01', 176868),
 ('19', 292329),
 ('12', 280052),
 ('09', 188304),
 ('11', 199185),
 ('18', 300438),
 ('23', 233095),
 ('08', 170462),
 ('06', 73137),
 ('13', 239953),
 ('07', 104624),
 ('03', 118617),
 ('04', 103641),
 ('15', 308633),
 ('10', 194347),
 ('02', 141190),
 ('21', 265082)]

In [12]:
offense_classification = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: ((x[6],x[7]),1))
sorted(offense_classification.reduceByKey(add).collect(),key=lambda x:x[1],reverse=True)
 

[(('341', 'PETIT LARCENY'), 822498),
 (('578', 'HARRASSMENT 2'), 604070),
 (('344', 'ASSAULT 3 & RELATED OFFENSES'), 521538),
 (('351', 'CRIMINAL MISCHIEF & RELATED OF'), 433358),
 (('109', 'GRAND LARCENY'), 429196),
 (('235', 'DANGEROUS DRUGS'), 285790),
 (('361', 'OFF. AGNST PUB ORD SENSBLTY &'), 283065),
 (('105', 'ROBBERY'), 198772),
 (('107', 'BURGLARY'), 191406),
 (('106', 'FELONY ASSAULT'), 184069),
 (('126', 'MISCELLANEOUS PENAL LAW'), 110468),
 (('110', 'GRAND LARCENY OF MOTOR VEHICLE'), 102061),
 (('359', 'OFFENSES AGAINST PUBLIC ADMINI'), 100118),
 (('347', 'INTOXICATED & IMPAIRED DRIVING'), 73730),
 (('236', 'DANGEROUS WEAPONS'), 73672),
 (('121', 'CRIMINAL MISCHIEF & RELATED OF'), 72416),
 (('352', 'CRIMINAL TRESPASS'), 66544),
 (('117', 'DANGEROUS DRUGS'), 62679),
 (('348', 'VEHICLE AND TRAFFIC LAWS'), 59140),
 (('112', 'THEFT-FRAUD'), 56762),
 (('118', 'DANGEROUS WEAPONS'), 50563),
 (('113', 'FORGERY'), 49303),
 (('233', 'SEX CRIMES'), 44117),
 (('340', 'FRAUDS'), 32054)

In [4]:
CMPLNT_FR_TM_min = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[2][3:5],1))
CMPLNT_FR_TM_min.reduceByKey(add).collect()

[('', 48),
 ('34', 10727),
 ('17', 11893),
 ('05', 102330),
 ('16', 10227),
 ('14', 11759),
 ('55', 99305),
 ('24', 11642),
 ('20', 222793),
 ('22', 11735),
 ('39', 10899),
 ('40', 208419),
 ('00', 1827614),
 ('26', 10487),
 ('01', 76567),
 ('57', 11133),
 ('19', 11233),
 ('52', 10961),
 ('44', 11092),
 ('12', 13552),
 ('09', 11421),
 ('11', 10260),
 ('18', 15547),
 ('49', 10670),
 ('37', 11309),
 ('28', 14297),
 ('41', 8806),
 ('56', 9972),
 ('23', 12350),
 ('LN', 1),
 ('43', 12130),
 ('59', 11952),
 ('36', 10723),
 ('42', 11591),
 ('08', 14653),
 ('06', 10769),
 ('50', 218622),
 ('13', 11963),
 ('32', 10863),
 ('07', 11486),
 ('54', 10326),
 ('35', 100300),
 ('29', 10600),
 ('03', 11612),
 ('46', 9998),
 ('38', 14474),
 ('53', 11092),
 ('45', 266329),
 ('31', 7881),
 ('30', 887332),
 ('33', 11058),
 ('04', 10516),
 ('15', 264658),
 ('48', 14856),
 ('27', 11455),
 ('10', 179026),
 ('25', 110572),
 ('47', 11553),
 ('02', 11506),
 ('58', 14566),
 ('51', 8521),
 ('21', 9200)]

In [5]:
offense_classification = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: ((x[6],x[7]),1))
sorted(offense_classification.reduceByKey(add).collect(),key=lambda x:x[0][0],reverse=True)

[(('KY_CD', 'OFNS_DESC'), 1),
 (('881', 'OTHER TRAFFIC INFRACTION'), 13),
 (('685', 'ADMINISTRATIVE CODES'), 18),
 (('678', ''), 223),
 (('678', 'MISCELLANEOUS PENAL LAW'), 8253),
 (('677', 'NYS LAWS-UNCLASSIFIED VIOLATION'), 6),
 (('677', 'OTHER STATE LAWS'), 233),
 (('676', 'NEW YORK CITY HEALTH CODE'), 31),
 (('675', 'ADMINISTRATIVE CODE'), 1446),
 (('675', ''), 4),
 (('672', 'LOITERING'), 20),
 (('578', 'HARRASSMENT 2'), 604070),
 (('578', ''), 71),
 (('577', 'UNDER THE INFLUENCE OF DRUGS'), 3),
 (('572', 'DISORDERLY CONDUCT'), 829),
 (('571', 'LOITERING/GAMBLING (CARDS, DIC'), 207),
 (('460', 'LOITERING/DEVIATE SEX'), 16),
 (('455', 'UNLAWFUL POSS. WEAP. ON SCHOOL'), 59),
 (('366', 'NEW YORK CITY HEALTH CODE'), 44),
 (('366', ''), 1),
 (('365', ''), 2),
 (('365', 'ADMINISTRATIVE CODE'), 9937),
 (('364', 'AGRICULTURE & MRKTS LAW-UNCLASSIFIED'), 83),
 (('364', ''), 402),
 (('364', 'OTHER STATE LAWS (NON PENAL LAW)'), 4),
 (('364', 'OTHER STATE LAWS (NON PENAL LA'), 5505),
 (('363', 

In [6]:
offense_classification = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: ((x[6],x[7]),1))
sorted(offense_classification.reduceByKey(add).collect(),key=lambda x:x[0][1],reverse=True)

[(('348', 'VEHICLE AND TRAFFIC LAWS'), 59140),
 (('455', 'UNLAWFUL POSS. WEAP. ON SCHOOL'), 59),
 (('577', 'UNDER THE INFLUENCE OF DRUGS'), 3),
 (('353', 'UNAUTHORIZED USE OF A VEHICLE'), 13826),
 (('112', 'THEFT-FRAUD'), 56762),
 (('343', 'THEFT OF SERVICES'), 2778),
 (('116', 'SEX CRIMES'), 10853),
 (('233', 'SEX CRIMES'), 44117),
 (('105', 'ROBBERY'), 198772),
 (('104', 'RAPE'), 13791),
 (('356', 'PROSTITUTION & RELATED OFFENSES'), 609),
 (('115', 'PROSTITUTION & RELATED OFFENSES'), 122),
 (('234', 'PROSTITUTION & RELATED OFFENSES'), 111),
 (('111', 'POSSESSION OF STOLEN PROPERTY'), 9112),
 (('232', 'POSSESSION OF STOLEN PROPERTY'), 20376),
 (('342', 'PETIT LARCENY OF MOTOR VEHICLE'), 908),
 (('341', 'PETIT LARCENY'), 822498),
 (('881', 'OTHER TRAFFIC INFRACTION'), 13),
 (('364', 'OTHER STATE LAWS (NON PENAL LAW)'), 4),
 (('364', 'OTHER STATE LAWS (NON PENAL LA'), 5505),
 (('677', 'OTHER STATE LAWS'), 233),
 (('343', 'OTHER OFFENSES RELATED TO THEF'), 9731),
 (('KY_CD', 'OFNS_DESC')

In [19]:
def CMPLNT_FR_DT_label(time):
	validity = None 
	base_type = 'DATETIME'
	semantic_type = 'Date of Occurrence'

	year = time[-4:]
	if year.isdigit():
		year = int(year)
	else:
		pass

	if time != "" and year in range(1900,2018):
		
		validity = "Valid"
		return (base_type,semantic_type,validity)
	else:
		validity = "Null"
		base_type = "Null"
		semantic_type = "Null"
		return (base_type,semantic_type,validity)

In [20]:
#header = crimedata.first() #header
#CMPLNT_FR_DT=crimedata.filter(lambda line: line != header)
CMPLNT_FR_DT = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[1],CMPLNT_FR_DT_label(x[1])))
CMPLNT_FR_DT.saveAsTextFile("output.out") 

In [9]:
def CMPLNT_FR_TM_label(time):
	validity = None 
	base_type = 'DATETIME'
	semantic_type = 'Time of occurrence'
	hour = time[:2] 

	if hour.isdigit():
		hour = int(hour)
	else:
		pass
	if time != "":
		
		validity = "Valid"
		return (base_type,semantic_type,validity)
	else:
		validity = "Null"
		base_type = "Null"
		semantic_type = "Null"
		return (base_type,semantic_type,validity)

In [10]:
CMPLNT_FR_TM = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[2],CMPLNT_FR_TM_label(x[2])))
CMPLNT_FR_TM.saveAsTextFile("output.out") 

In [29]:
def CMPLNT_TO_DT_label(time):
	validity = None 
	base_type = 'DATETIME'
	semantic_type = 'Ending Date of Occurrence'

	year = time[-4:]
	if year.isdigit():
		year = int(year)
	else:
		pass

	if time != "" and year in range(1900,2018):
		validity = "Valid"
		return (base_type,semantic_type,validity)
	elif time == "":
		validity = "Null"
		base_type = "Null"
		semantic_type = "Null"
		return (base_type,semantic_type,validity) 
	else:
		validity = "Invalid"
		base_type = "Null"
		semantic_type = "Null"
		return (base_type,semantic_type,validity)

In [30]:
CMPLNT_TO_DT = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[3],CMPLNT_TO_DT_label(x[3])))
CMPLNT_TO_DT.saveAsTextFile("output.out")

In [34]:
def CMPLNT_TO_TM_label(time):
	validity = None 
	base_type = 'DATETIME'
	semantic_type = 'Ending Time of Occurrence'
	hour = time[:2] 
	
	if hour.isdigit():
		hour = int(hour)
	else:
		pass
	if time != "" and hour in range(0,25):
		
		validity = "Valid"
		return (base_type,semantic_type,validity)

	elif time == "":
		validity = "Null"
		base_type = "Null"
		semantic_type = "Null" 
		return (base_type,semantic_type,validity)
	else:
		validity = "Invalid"
		base_type = "Null"
		semantic_type = "Null"
		return (base_type,semantic_type,validity)

In [36]:
CMPLNT_TO_TM = crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[4],CMPLNT_TO_TM_label(x[4])))
CMPLNT_TO_TM.saveAsTextFile("output.out") 

In [43]:
def RPT_DT_label(time):
	validity = None 
	base_type = 'DATETIME'
	semantic_type = 'Date event was reported to police '

	if time == "":
		validity = "Null"
		base_type = "Null"
		semantic_type = "Null"
		return (base_type,semantic_type,validity)
	elif re.match("(\d{2})[/](\d{2})[/](\d{4})$",time) != None:
		month  = int(time.split('/')[0])
		date   = int(time.split('/')[1])
		year   = int(time.split('/')[2])

		if month in range(1,13) and date in range(1,32) and year in range(1900,2018):
			validity = "Valid"
			return (base_type,semantic_type,validity)
		else:
			validity = "Invalid"
			return (base_type,semantic_type,validity)
	else:
		validity = "Invalid"
		return (base_type,semantic_type,validity)

In [44]:
header = crimedata.first() #header
RPT_DT=crimedata.filter(lambda line: line != header)
RPT_DT = RPT_DT.mapPartitions(lambda x: reader(x)).map(lambda x: (x[5],RPT_DT_label(x[5])))
RPT_DT.saveAsTextFile("output.out")

In [46]:
crimedata.mapPartitions(lambda x: reader(x)).map(lambda x: (x[6],1)).reduceByKey(add).collect()


[('362', 3),
 ('349', 47),
 ('104', 13791),
 ('109', 429205),
 ('114', 13984),
 ('572', 829),
 ('571', 207),
 ('357', 11),
 ('343', 12511),
 ('101', 4574),
 ('238', 2645),
 ('237', 168),
 ('110', 102061),
 ('354', 95),
 ('675', 1450),
 ('119', 48),
 ('678', 8476),
 ('352', 66549),
 ('231', 2477),
 ('355', 12341),
 ('115', 122),
 ('102', 93),
 ('364', 5994),
 ('103', 33),
 ('676', 31),
 ('577', 3),
 ('347', 73733),
 ('363', 810),
 ('230', 300),
 ('107', 191407),
 ('122', 122),
 ('235', 285801),
 ('685', 18),
 ('106', 184124),
 ('233', 44130),
 ('112', 56768),
 ('672', 20),
 ('345', 1468),
 ('126', 112266),
 ('348', 59142),
 ('359', 100134),
 ('455', 59),
 ('236', 73680),
 ('342', 908),
 ('111', 9113),
 ('341', 822508),
 ('460', 16),
 ('353', 13826),
 ('124', 2305),
 ('118', 50563),
 ('356', 618),
 ('358', 17739),
 ('578', 604141),
 ('120', 389),
 ('113', 49304),
 ('346', 750),
 ('105', 198774),
 ('125', 3756),
 ('365', 9939),
 ('340', 32057),
 ('361', 283102),
 ('121', 72421),
 ('116', 

In [49]:
def KY_CD_Label(code):
	validity = None 
	base_type = 'INT'
	semantic_type = 'Offense Classification Code'

	if code == "" or len(code) == 0:
		validity = 'Null'
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)

	elif len(code) > 3:
		validity = 'Invalid'
		base_type = 'INT'
		semantic_type = 'Offense Classification Code'
		return (base_type,semantic_type,validity)

	else:
		validity = "Valid"
		return (base_type,semantic_type,validity)

In [51]:
header = crimedata.first() #header
KY_CD=crimedata.filter(lambda line: line != header)
KY_CD = KY_CD.mapPartitions(lambda x: reader(x)).map(lambda x: (x[6],KY_CD_Label(x[6])))
KY_CD.saveAsTextFile("output.out")

In [3]:
def OFNS_DESC_label(text):
	validity = None 
	base_type = 'TEXT'
	semantic_type = 'Description of Offense'

	if len(text) == 0 or text == "":
		validity = 'Null' 
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)

	else:
		validity = 'Valid' 
		base_type = 'TEXT'
		semantic_type = 'Description of Offense'
		return (base_type,semantic_type,validity)

In [4]:
header = crimedata.first() #header
OFNS_DESC=crimedata.filter(lambda line: line != header)
OFNS_DESC = OFNS_DESC.mapPartitions(lambda x: reader(x)).map(lambda x: (x[7],OFNS_DESC_label(x[7])))
OFNS_DESC.saveAsTextFile("output.out")

In [10]:
def PD_CD_Label(code):
	validity = None 
	base_type = 'INT'
	semantic_type = 'Internal Classification Code'

	if code == "":
		validity = 'Null'
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)

	else:
		code = int(code)
		if code not in range(1000):
			validity = 'Invalid'
			base_type = 'INT'
			semantic_type = 'Offense Classification Code'
			return (base_type,semantic_type,validity)
		else:
			validity = 'Valid'
			return (base_type,semantic_type,validity)

In [11]:
header = crimedata.first() #header
PD_CD=crimedata.filter(lambda line: line != header)
PD_CD = PD_CD.mapPartitions(lambda x: reader(x)).map(lambda x: (x[8],PD_CD_Label(x[8])))
PD_CD.saveAsTextFile("output.out")

In [14]:
def PD_DESC_label(text):
	validity = None 
	base_type = 'TEXT'
	semantic_type = 'Description of Internal Classification'

	if len(text) == 0 or text == "":
		validity = 'Null' 
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)

	else:
		validity = 'Valid' 
		base_type = 'TEXT'
		semantic_type = 'Description of Offense'
		return (base_type,semantic_type,validity)

In [15]:
header = crimedata.first() #header
PD_DESC = crimedata.filter(lambda line: line != header)
PD_DESC = PD_DESC.mapPartitions(lambda x: reader(x)).map(lambda x: (x[9],PD_DESC_label(x[9])))
PD_DESC.saveAsTextFile("output.out")

In [16]:
def CRM_ATPT_CPTD_CD_label(indicator):
	validity = None 
	base_type = 'TEXT'
	semantic_type = 'Indicator of completeness' 

	indicator_type = ['COMPLETED','ATTEMPTED']

	if indicator == "" or len(indicator) == 0:
		validity = 'Null' 
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)

	elif indicator in indicator_type:
		validity = 'Valid'
		return (base_type,semantic_type,validity)

	else:
		validity = 'Invalid'
		return (base_type,semantic_type,validity)

In [17]:
header = crimedata.first() #header
CRM_ATPT_CPTD_CD = crimedata.filter(lambda line: line != header)
CRM_ATPT_CPTD_CD = CRM_ATPT_CPTD_CD.mapPartitions(lambda x: reader(x)).map(lambda x: (x[10],CRM_ATPT_CPTD_CD_label(x[10])))
CRM_ATPT_CPTD_CD.saveAsTextFile("output.out")

In [18]:
def LAW_CAT_CD_label(level):
	validity = None 
	base_type = 'TEXT'
	semantic_type = 'Level of Offense' 
	level_of_offense = ['FELONY', 'MISDEMEANOR', 'VIOLATION']

	if level == "" or len(level) == 0:
		validity = 'Null' 
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)
	elif level not in level_of_offense:
		validity = 'Invalid'
		return (base_type,semantic_type,validity)
	else:
		validity = 'Valid'
		return (base_type,semantic_type,validity)

In [19]:
header = crimedata.first() #header
LAW_CAT_CD = crimedata.filter(lambda line: line != header)
LAW_CAT_CD = LAW_CAT_CD.mapPartitions(lambda x: reader(x)).map(lambda x: (x[11],LAW_CAT_CD_label(x[11])))
LAW_CAT_CD.saveAsTextFile("output.out")

In [47]:
def BORO_NM_label(borough,precinct):
	
	boro_precinct = {'BRONX':list(range(40,51))+[52], 'QUEENS':list(range(100,116)), \
	'MANHATTAN':[1,5,6,7,9,10,13,17,18,19,20,22,23,24,25,26,28,30,32,33,34],\
	'BROOKLYN':[60,61,62,63,66,67,68,69,70,71,72,73,75,76,77,78,79,81,83,84,88,90,94],\
	'STATEN ISLAND':list(range(120,124))}

	validity = None 
	base_type = 'TEXT'
	semantic_type = 'Borough Where Incident Occurred'
	if precinct != "":
		precinct = int(precinct) 

	if borough == "" or len(borough) == 0:
		validity = 'Null' 
		base_type = 'Null'
		semantic_type = 'Null'
		return (base_type,semantic_type,validity)

	elif precinct not in boro_precinct[borough]:
		validity = 'Invalid'
		return (base_type,semantic_type,validity)
	else:
		validity = 'Valid'
		return (base_type,semantic_type,validity)

In [48]:
header = crimedata.first() #header
BORO_NM = crimedata.filter(lambda line: line != header)
BORO_NM = BORO_NM.mapPartitions(lambda x: reader(x)).map(lambda x: (x[13],BORO_NM_label(x[13],x[14])))
BORO_NM.saveAsTextFile("output.out")